libraries/azure_backend.rb

```diff
@@ -0,0 +1,788 @@
+  # @param singular [TrueClass, FalseClass] Define whether or not the expected result is for a singular resource (default - true).
+  def validate_short_desc(resource_list, filter, singular = true)
+    message = "#{@__resource_name__}: #{@display_name}. Unable to get the resource short description with the provided data: #{filter}"
+    if resource_list.nil?
+      resource_fail(message)
+      false
+    elsif resource_list.empty?
+      empty_response_warn(message)
+      false
+    elsif singular && resource_list.size > 1
+      resource_fail
+      false
+    else
+      true
+    end
+  end
+
+  # Get the paginated result.
+  # The next_link url won't be validated since it is provided by the Azure Rest API.
+  # @see https://docs.microsoft.com/en-us/rest/api/azure/#async-operations-throttling-and-paging
+  #
+  # @param next_link [String] The nextLink url provided by the Azure Rest API.
+  # @return [Hash] The HTTP response body in JSON/Hash format with symbolized keys.
+  #
+  def get_next_link(next_link)
+    @azure.rest_get_call(next_link)
+  end
+
+  # Enforce specific resource type constraint in static resources, e.g: 'azure_virtual_machine'.
+  #
+  # @return [String] The resource type, 'Microsoft.Compute/virtualMachines'.
+  # @param resource_provider [String] The resource type, 'Microsoft.Compute/virtualMachines'.
+  # @param opts [Hash] Parameters to check if the ':resource_provider' key exists.
+  # If the resource type exists in the opts, this will raise an ArgumentError.
+  # The resource_provider parameter should not be provided by the user in static resources.
+  # Otherwise, static resource would not behave as expected if a different resource type is provided.
+  #
+  def specific_resource_constraint(resource_provider, opts)
+    if opts.is_a?(Hash)
+      if opts.key?(:resource_provider)
+        raise ArgumentError, "#{@__resource_name__}: The `resource_provider` parameter is not allowed."\
+          " `#{resource_provider}` is predefined for this resource."
+      end
+    else
+      raise ArgumentError, "#{@__resource_name__}: Parameters must be provided in an Hash object."
+    end
+    resource_provider
+  end
+
+  # Intercept failed resource queries.
+  #
+  # Inform if it is an api issue and present the suggested version by the Azure Rest API.
+  #
+  # This should be used to ensure to fail the resources properly if they can not be created.
+  def catch_failed_resource_queries
+    yield
+  # Inform user if it is an API incompatibility issue and recommend how to solve it.
+  rescue UnsuccessfulAPIQuery::UnexpectedHTTPResponse::InvalidApiVersionParameter => e
+    api_version_suggested_list = e.suggested_api_version
+    message = "Incompatible api version is provided.\n"\
+    "The list of api versions suggested by the Azure REST API is #{api_version_suggested_list}.\n #{e.message}"\
+    'Note that this list includes the invalid api version and it should be removed before using the list.'
```

Comment:
Should this say `'Note that if this list`?

Reply:
The list is created by `rescue UnsuccessfulAPIQuery::UnexpectedHTTPResponse::InvalidApiVersionParameter => e` and it will include the wrong API version. However, it is safer to use your suggestion.